src/plugins/chatMarkers/ChatMarkersPlugin.ts

```diff
@@ -0,0 +1,466 @@
+      this.pluginAPI.getConnection().registerHandler(this.onChatMarkersMessage, null, 'message');
+
+      this.pluginAPI.registerChatWindowInitializedHook((chatWindow: ChatWindow, contact: Contact) => {
+         this.chatWindowInitializedHook(chatWindow, contact);
+      });
+   }
+
+   private supportsChatMarkers(jid: IJID): Promise<boolean> {
+      if (jid.isBare()) {
+         // if bare JID, sender MAY send chat markers
+         this.pluginAPI.Log.debug(`${jid.full} supports chat markers. Yaay! =)`);
+         return Promise.resolve(true);
+      } else {
+         // if full JID, sender SHOULD try to determine if recipient supports chat markers
+         return this.pluginAPI.getDiscoInfoRepository().hasFeature(jid, [Namespace.get(CHATMARKERS)]).then((hasFeature) => {
+            if (hasFeature) {
+               this.pluginAPI.Log.debug(`${jid.full} supports chat markers. Yaay! =)`);
+            } else {
+               this.pluginAPI.Log.debug(`${jid.full} doesn't support chat markers. =(`);
+            }
+            return hasFeature;
+         }).catch(() => {
+            this.pluginAPI.Log.debug(`${jid.full} doesn't support chat markers. =(`);
+            return false;
+         });
+      }
+   }
+
+   // add "markable" element according to XEP-0333
+   private addMarkable(xmlStanza: Strophe.Builder, msg: Message) {
+      this.pluginAPI.Log.debug(`adding ${MARKABLE} tag. Yaay! =)`);
+      xmlStanza.c(MARKABLE, {
+         xmlns: Namespace.get(CHATMARKERS)
+      }).up();
+   }
+
+   // send "received" message according to XEP-0333
+   private sendReceived(lastReceivedMsgId: string, to: IJID) {
+      this.pluginAPI.Log.debug(`sending ${RECEIVED} message. Yaay! =)`);
+      this.pluginAPI.send($msg({
+         to: to.full
+      }).c(RECEIVED, {
+         xmlns: Namespace.get(CHATMARKERS),
+         id: lastReceivedMsgId
+      }));
+   }
+
+   // send "displayed" message according to XEP-0333
+   private sendDisplayed(lastDisplayedMsgId: string, to: IJID) {
+      this.pluginAPI.Log.debug(`sending ${DISPLAYED} message. Yaay! =)`);
+      this.pluginAPI.send($msg({
+         to: to.full
+      }).c(DISPLAYED, {
+         xmlns: Namespace.get(CHATMARKERS),
+         id: lastDisplayedMsgId
+      }));
+   }
+
+   // send "acknowledged" message according to XEP-0333
+   private sendAcknowledged(lastAcknowledgedMsgId: string, to: IJID) {
+      this.pluginAPI.Log.debug(`sending ${ACKNOWLEDGED} message. Yaay! =)`);
+      this.pluginAPI.send($msg({
+         to: to.full
+      }).c(ACKNOWLEDGED, {
+         xmlns: Namespace.get(CHATMARKERS),
+         id: lastAcknowledgedMsgId
+      }));
+   }
+
+   private setMessagesChatMarkersAcknowledged(lastAcknowledgedMsgUid: string, from: IJID) {
+      this.pluginAPI.Log.debug(`setting previous messages ${ACKNOWLEDGED}. Yaay! =)`);
+      let contact = this.pluginAPI.getContact(from);
+      if (!!contact) {
```

Comment:
It's not required to have two exclamation marks here, but it would reduce the indentation of the following code if you return if there is no contact.